tests/zfs-tests/tests/functional/deadman/deadman_ratelimit.ksh

```diff
@@ -0,0 +1,76 @@
+#!/bin/ksh -p
+#
+# CDDL HEADER START
+#
+# The contents of this file are subject to the terms of the
+# Common Development and Distribution License (the "License").
+# You may not use this file except in compliance with the License.
+#
+# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+# or http://www.opensolaris.org/os/licensing.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Portions Copyright 2021 iXsystems, Inc.
+#
+
+# DESCRIPTION:
+#	Verify spa deadman events are rate limited
+#
+# STRATEGY:
+#	1. Reduce the zfs_slow_io_events_per_second to 1.
+#	2. Reduce the zfs_deadman_ziotime_ms to 1ms.
+#	3. Write data to a pool and read it back.
+#	4. Verify deadman events have been produced at a reasonable rate.
+#
+
+. $STF_SUITE/include/libtest.shlib
+. $STF_SUITE/tests/functional/deadman/deadman.cfg
+
+verify_runnable "both"
+
+function cleanup
+{
+	default_cleanup_noexit
+
+	set_tunable64 SLOW_IO_EVENTS_PER_SECOND $OLD_SLOW_IO_EVENTS
+	set_tunable64 DEADMAN_ZIOTIME_MS $ZIOTIME_DEFAULT
+}
+
+log_assert "Verify spa deadman events are rate limited"
+log_onexit cleanup
+
+OLD_SLOW_IO_EVENTS=$(get_tunable SLOW_IO_EVENTS_PER_SECOND)
+log_must set_tunable64 SLOW_IO_EVENTS_PER_SECOND 1
+log_must set_tunable64 DEADMAN_ZIOTIME_MS 1
+
+# Create a new pool in order to use the updated deadman settings.
+default_setup_noexit $DISK1
+log_must zpool events -c
+
+mntpnt=$(get_prop mountpoint $TESTPOOL/$TESTFS)
+log_must file_write -b 1048576 -c 8 -o create -d 0 -f $mntpnt/file
+log_must zinject -a
+log_must dd if=$mntpnt/file of=$TEST_BASE_DIR/devnull oflag=sync
+
+sleep 10
```

Comment:
Is the sleep here needed with such a short deadman timeout?

Reply:
I'm not sure what's going on here. I put the sleep there under the understanding that events might not have finished making their way into the queue by the time we go to look for them. Ten seconds should have been overkill, but apparently it's not even enough in some cases. There are some test bots failing to see any deadman events somehow.

Unfortunately, I can't reproduce these failures locally. The tests are working as intended on FreeBSD and Linux in my dev environments. I'll continue auditing the zevent code. Perhaps there are some more hidden issues like the ratelimit_dropped one I came across already.

Reply:
I'd suggest updating the test to use the `file_wait_event` function from `events_common.ksh`.  That checks the ZED log for the message and you can set a maximum timeout...  although I guess that wouldn't be portable for FreeBSD.  But you could write something similar.

One other thing you might try is to replace the `zinject -a` with a `zpool export; zpool import`.  While `zinject -a` should drop the cache, it may not drop everything if for example the buffers are still being written back.  In which case they'll be cache hits for your dd read and you won't trigger the deadman probably.